polkadot/node/tracking-allocator/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Tracking/limiting global allocator. Calculates the peak allocation between two checkpoints for
+//! the whole process. Accepts an optional limit and a failure handler which is called if the limit
+//! is overflown.
+
+use core::alloc::{GlobalAlloc, Layout};
+use std::{
+	ptr::null_mut,
+	sync::atomic::{AtomicBool, Ordering},
+};
+
+struct TrackingAllocatorData {
+	lock: AtomicBool,
+	current: isize,
+	peak: isize,
+	limit: isize,
+	failure_handler: Option<Box<dyn Fn()>>,
+}
+
+impl TrackingAllocatorData {
+	#[inline]
+	fn lock(&self) {
+		loop {
+			// Try to acquire the lock.
+			if self
+				.lock
+				.compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+				.is_ok()
+			{
+				break
+			}
+			// We failed to acquire the lock; wait until it's unlocked.
+			//
+			// In theory this should result in less coherency traffic as unlike `compare_exchange`
+			// it is a read-only operation, so multiple cores can execute it simultaneously
+			// without taking an exclusive lock over the cache line.
+			while self.lock.load(Ordering::Relaxed) {
+				std::hint::spin_loop();
+			}
+		}
+	}
+
+	#[inline]
+	fn unlock(&self) {
+		self.lock.store(false, Ordering::Release);
+	}
+
+	fn start_tracking(&mut self, limit: isize, failure_handler: Option<Box<dyn Fn()>>) {
+		self.lock();
+		self.current = 0;
+		self.peak = 0;
+		self.limit = limit;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = self.failure_handler.take();
+		self.failure_handler = failure_handler;
+		self.unlock();
+		core::mem::drop(old_handler);
+	}
+
+	fn end_tracking(&mut self) -> isize {
+		self.lock();
+		let peak = self.peak;
+		self.limit = 0;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = self.failure_handler.take();
+		self.unlock();
+		core::mem::drop(old_handler);
+		peak
+	}
+
+	#[inline]
+	fn track(&mut self, alloc: isize) -> bool {
+		self.lock();
+		self.current += alloc;
+		if self.current > self.peak {
+			self.peak = self.current;
+		}
+		let within_limits = self.limit == 0 || self.peak <= self.limit;
```

Comment:
If we'd set `limit` to `isize::MAX` by default instead of `0` then the `self.limit == 0` could be removed, saving one branch.

(Probably won't make any difference though.)